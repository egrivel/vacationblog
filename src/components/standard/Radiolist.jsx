import React from 'react';
import PropTypes from 'prop-types';

import Formrow from './Formrow';

class Radiolist extends React.Component {
  static propTypes = {
    fieldId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    list: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })).isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string
  }

  static defaultProps = {
    value: ''
  }

  render() {
    const {fieldId, label, list, onChange, value} = this.props;

    const options = [];
    for (let i = 0; i < list.length; i++) {
      const item = list[i];
      let checked = false;
      if (item.value === value) {
        checked = true;
      }
      options.push(
        <label
          className="radiobutton"
          htmlFor={fieldId + '-' + i}
          key={'r-' + i}
          onClick={event => onChange(event.target.value, fieldId)}
        >
          <input
            id={fieldId + '-' + i}
            name={fieldId}
            type="radio"
            value={item.value}
            checked={checked}
            onClick={event => onChange(event.target.value, fieldId)}
          />
          {item.label + ' '}
        </label>
      );
    }
    return (
      <Formrow
        key={'k-' + fieldId}
        label={label}
        labelFor={fieldId}
      >
        <span className="radioList">
          {options}
        </span>
      </Formrow>
    );
  }
}

export default Radiolist;
