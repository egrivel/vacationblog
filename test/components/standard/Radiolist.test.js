import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import enzyme from 'enzyme';

import Radiolist from '../../../src/components/standard/Radiolist';

/*
 * Test strategy
 * The component has one optional prop (value) and one callback
 * function (onChange). The requirements are:
 *  - the component renders with or without a value
 *  - the onChange is fired when changing the selected item, either by
 *    clicking on the button or on the label
 */
describe('components/standard/Radiolist', () => {
  const testFieldId = 'test-field-id';
  const testLabel = 'Test Label';
  const testItemLabel1 = 'Test Item Label 1';
  const testItemValue1 = 'test-item-value-1';
  const testItemLabel2 = 'Test Item Label 2';
  const testItemValue2 = 'test-item-value-2';
  const testItemLabel3 = 'Test Item Label 3';
  const testItemValue3 = 'test-item-value-3';
  let onChangeStub;
  let testEvent;
  let testList;

  beforeEach(() => {
    onChangeStub = sinon.stub();
    testEvent = {
      preventDefault() {},
      target: {}
    };
    testList = [
      {
        label: testItemLabel1,
        value: testItemValue1
      }, {
        label: testItemLabel2,
        value: testItemValue2
      }, {
        label: testItemLabel3,
        value: testItemValue3
      }
    ];
  });

  afterEach(() => {
    sinon.restore();
  });

  describe('#render', () => {
    test('renders without a value', () => {
      const render = renderer.create(
        <Radiolist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });

    test('renders checked', () => {
      const render = renderer.create(
        <Radiolist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
          value={testItemValue2}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });
  });

  describe('handles onChange', () => {
    it('clicking on the input passes the selected value back', () => {
      const component = enzyme.shallow(
        <Radiolist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
        />
      );
      testEvent.target.value = testItemValue1;
      const inputList = component.find('input');
      for (let i = 0; i < inputList.length; i++) {
        if (inputList.at(i).props().value === testItemValue1) {
          inputList.at(i).simulate('click', testEvent);
        }
      }
      expect(onChangeStub.callCount).toBe(1);
      expect(onChangeStub.args[0][0]).toBe(testItemValue1);
      expect(onChangeStub.args[0][1]).toBe(testFieldId);
    });

    it('clicking on the label passes the selected value back', () => {
      const component = enzyme.shallow(
        <Radiolist
          fieldId={testFieldId}
          label={testLabel}
          list={testList}
          onChange={onChangeStub}
        />
      );
      testEvent.target.value = testItemValue1;
      // Find the input, then click on the corresponding label
      const inputList = component.find('input');
      const labelList = component.find('label');
      for (let i = 0; i < inputList.length; i++) {
        if (inputList.at(i).props().value === testItemValue1) {
          labelList.at(i).simulate('click', testEvent);
        }
      }
      expect(onChangeStub.callCount).toBe(1);
      expect(onChangeStub.args[0][0]).toBe(testItemValue1);
      expect(onChangeStub.args[0][1]).toBe(testFieldId);
    });
  });
});
