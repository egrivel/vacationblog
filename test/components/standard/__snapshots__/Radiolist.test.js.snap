// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/standard/Radiolist #render renders checked 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <span
      className="radioList"
    >
      <label
        className="radiobutton"
        htmlFor="test-field-id-0"
        onClick={[Function]}
      >
        <input
          checked={false}
          id="test-field-id-0"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-1"
        />
        Test Item Label 1 
      </label>
      <label
        className="radiobutton"
        htmlFor="test-field-id-1"
        onClick={[Function]}
      >
        <input
          checked={true}
          id="test-field-id-1"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-2"
        />
        Test Item Label 2 
      </label>
      <label
        className="radiobutton"
        htmlFor="test-field-id-2"
        onClick={[Function]}
      >
        <input
          checked={false}
          id="test-field-id-2"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-3"
        />
        Test Item Label 3 
      </label>
    </span>
  </div>
</div>
`;

exports[`components/standard/Radiolist #render renders without a value 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <span
      className="radioList"
    >
      <label
        className="radiobutton"
        htmlFor="test-field-id-0"
        onClick={[Function]}
      >
        <input
          checked={false}
          id="test-field-id-0"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-1"
        />
        Test Item Label 1 
      </label>
      <label
        className="radiobutton"
        htmlFor="test-field-id-1"
        onClick={[Function]}
      >
        <input
          checked={false}
          id="test-field-id-1"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-2"
        />
        Test Item Label 2 
      </label>
      <label
        className="radiobutton"
        htmlFor="test-field-id-2"
        onClick={[Function]}
      >
        <input
          checked={false}
          id="test-field-id-2"
          name="test-field-id"
          onClick={[Function]}
          type="radio"
          value="test-item-value-3"
        />
        Test Item Label 3 
      </label>
    </span>
  </div>
</div>
`;
